src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageSearch/IPackageSearchResultRenderer.cs

```diff
@@ -28,11 +28,17 @@ internal interface IPackageSearchResultRenderer
         /// <param name="source">The source</param>
         /// <param name="error">The error message to be rendered</param>
         /// <returns></returns>
-        void Add(PackageSource source, string error);
+        void Add(PackageSource source, PackageSearchProblem packageSearchProblem);
 
         /// <summary>
         /// Finishes the rendering operation.
         /// </summary>
         void Finish();
+
+        /// <summary>
+        ///  Renders a problem that prevented the search from happening at all.
+        /// </summary>
+        /// <param name="packageSearchProblem"></param>
+        void RenderProblem(PackageSearchProblem packageSearchProblem);
```

Comment:
There is two problem sections: in the general section, in the source section.

For example, in this JSON output, there is two problems properties. One of them is found as part of the general properties. And the second one is found as part of the "source Name1" properties. If there are any problems that makes the searching operation not-proceed, it will be added in the first problems section. However, if there are problems that originate from a specific source, that will be added to the problems section that is found under the specific source. For example, let's say the search service for `source Name1` is not available, that problem will be added in the problems section for `source Name1`. This `RenderProblem` will add problems that are originate from a specific source. the method `Finish` on the other hand is called once iterating through all the sources is over. As a result, it cannot add per source specific problems to the JSON efficiently.


      {
        "version": 1,
        "problems": [
          {
            "level": "error",
            "text": "An error occurred!"
          }
        ],
        "searchResult": [
          {
            "source": "source Name1",
            "problems": [
              {
                "level": "error",
                "text": "An error occurred!"
              }
            ],
            "packages": [
              {
                "Owners": "",
                "totalDownloads": "",
                "latestVersion": "",
                "id": ""
              }
            ]
          } ]
          }
        ]
      }

Reply:
I see. 
So we have 2 classes of problems. I think the name should still be add problem, and not render immediately though.